### DISCUSSION TITLE
Efficiency of a Query

### DISCUSSION THREAD

**Discussion by Tonaion02** [bevyengine/bevy#15069] [Q&A]
The efficiency of a Query depends on the amount of Component that i establish?
So for example, the retrieving of entities with one component is more efficiently than retrieving entities with three or four components?

**Comment by NthTensor** [comment]
Nope. Bevy sorts data into tables called "archetypes" which each column is a component and each row is an entity. All entities with the same exact set of components live in the same archetype table.

Accessing a single entity is basically `O(1)`, it's just a table lookup.

Iteration is slightly more complex, because query can match multiple archetype tables. Iterating through a single table is very efficient, but there is a cost to jumping from one table to the next. So in practice, the main thing that will slow down iteration is *fragmentation*. Having lots of different entities with a different set of components on each will force you to jump through a ton of archetype tables.

I wouldn't worry about this unless you start actually hitting performance problems, but in general it's the number of different combinations of components you have to watch out for, rather than the number of entities or the number of components on those entities.

**Reply by NotAFile** [reply]
More succinctly: the cost of a query scales roughly with the number of archetypes matched. Thus paradoxically, in general a query is more efficient the more specific it is.

In effect this means: don't worry about query performance. Just query for what your system needs and it will be as fast as it can be.
